### PR TITLE
Categorise test cases into test suite(s)

### LINKED ISSUE #245: Categorise test cases into test suite(s)
Thank you for the great extension!

I think the vscode-meson should categorise test cases into test suite(s), if exists.

Steps:

1. I am using vscode-meson v1.24.0 with a C/C++ project (gtk)
2. Select VSCode's "Testing" feature from the Activity Bar
3. Observe a long list of test cases. The gtk project categorises the Meson test cases into Meson test suite(s). I think vscode&#8209;meson should categorise too.

Screenshot 2024-06-30 121436

### PR DESCRIPTION
**Pull request by cskeogh**
Meson test cases may belong to test suite(s). If so, in VSCode's "Testing" feature from the Activity Bar, add the test suite(s) as a parent item and add the test cases as children.

The user has the choice to run all tests belonging to a test suite. Will pass the test suite to the meson test command --suite. The option --suite is needed to run test cases that the meson config purposely excludes (for example, test suite "flaky").

Fixes #245

Quote the "test case" in the meson command, so can run test cases with a space in the name. The quotes are needed because `cp.ExecFileOptions` sets `shell=true`.

Fixes #256

### CODE REVIEW

**Comment by cskeogh** [comment]
Using the gtk project, showing the test suite(s):

Screenshot 2024-06-30 124213

**Comment by tristan957** [comment]
Thanks for the PR. I will try to test and review it this week.

File: src/tests.ts [hunk 1]
```diff
@@ -4,18 +4,73 @@ import { Tests, DebugEnvironmentConfiguration } from "./types";
...
+}
 
-  controller.items.forEach((item) => {
-    if (!tests.some((test) => item.id == test.name)) {
-      controller.items.delete(item.id);
+function deleteTestsFromControllerNotVisited(
+  controller: vscode.TestController,
+  testsVisited: WeakMap<vscode.TestItem, boolean>,
+) {
+  for (const [test_id, test] of controller.items) {
```
Comment (tristan957) [inline src/tests.ts:53]: Can you make variable names camelCase?

File: src/tests.ts [hunk 2]
```diff
@@ -4,18 +4,73 @@ import { Tests, DebugEnvironmentConfiguration } from "./types";
...
     }
-  });
+  }
+}
 
-  for (let testDescr of tests) {
-    let testItem = controller.createTestItem(testDescr.name, testDescr.name);
-    controller.items.add(testItem);
+export async function rebuildTests(controller: vscode.TestController) {
+  let tests = await getMesonTests(workspaceState.get<string>("mesonbuild.buildDir")!);
```
Comment (tristan957) [inline src/tests.ts:70]: const

**Comment by tristan957** [comment]
What do you think of #237? Should I merge that one first?

**Comment by cskeogh** [comment]
Updated the PR to fix the review comments.

Note the PR does have some limitations:

1. Gtk's test case "about state" is in 2 test suites "gtk+:a11y" and "gtk+:flaky". This PR shows "about state" correctly in both test suites, but they are separate instances of `TestItem`. Ideally, "about state" would be one instance of `TestItem`, so that the failure icon shows on both (see screenshot yellow highlight). But VSCode doesn't support one `TestItem` with 2 parents. I tried.
 Screenshot 2024-07-02 215050
 Screenshot 2024-07-02 215210

1.  The user can run the test suite, which will correctly run all test cases belonging to the test suite. But the children `TestItem` do not show the success / failure icons (see screenshot yellow highlight). Only the test suite shows success / failure. In the below example, I ran test suite "gtk+:gdk" and 7 test cases passed and 1 failed:
Screenshot 2024-07-02 220438
  ```
Ok:                 7   
Expected Fail:      0   
Fail:               1   
Unexpected Pass:    0   
Skipped:            0   
Timeout:            0 
   ```

**Comment by tristan957** [comment]
Thanks for explaining the limitations. I see that there is this function. Do we have access to the test runs, so if someone runs a suite, we can properly annotate the test? https://vscode-api.js.org/interfaces/vscode.TestRun.html#passed

**Comment by cskeogh** [comment]
I'm working on resolving the limitations. Closing this PR until I resolve.

**Comment by tristan957** [comment]
Alright. Really appreciate you looking into this more!

**Pull request closed without merging**